Firestore/core/src/firebase/firestore/api/query_core.mm

```diff
@@ -0,0 +1,327 @@
+            FirestoreErrorCode::Unavailable,
+            "Failed to get documents from server. (However, these documents "
+            "may exist in the local cache. Run again without setting source to "
+            "FirestoreSourceServer to retrieve the cached documents.)"});
+      } else {
+        listener_->OnEvent(std::move(snapshot));
+      }
+    };
+
+    void Resolve(ListenerRegistration&& registration) {
+      registration_promise_.set_value(std::move(registration));
+    }
+
+   private:
+    Source source_;
+    QuerySnapshot::Listener listener_;
+
+    std::promise<ListenerRegistration> registration_promise_;
+  };
+
+  auto listener = absl::make_unique<ListenOnce>(source, std::move(callback));
+  auto listener_unowned = listener.get();
+
+  ListenerRegistration registration =
+      AddSnapshotListener(std::move(options), std::move(listener));
+
+  listener_unowned->Resolve(std::move(registration));
+}
+
+ListenerRegistration Query::AddSnapshotListener(
+    ListenOptions options, QuerySnapshot::Listener&& user_listener) {
+  // Convert from ViewSnapshots to QuerySnapshots.
+  class Converter : public EventListener<ViewSnapshot> {
+   public:
+    Converter(Query* parent, QuerySnapshot::Listener&& user_listener)
+        : firestore_(parent->firestore()),
+          query_(parent->query()),
+          user_listener_(std::move(user_listener)) {
+    }
+
+    void OnEvent(StatusOr<ViewSnapshot> maybe_snapshot) override {
+      if (!maybe_snapshot.status().ok()) {
+        user_listener_->OnEvent(maybe_snapshot.status());
+        return;
+      }
+
+      ViewSnapshot snapshot = std::move(maybe_snapshot).ValueOrDie();
+      SnapshotMetadata metadata(snapshot.has_pending_writes(),
+                                snapshot.from_cache());
+
+      QuerySnapshot result(firestore_, query_, std::move(snapshot),
+                           std::move(metadata));
+
+      user_listener_->OnEvent(result);
+    }
+
+   private:
+    std::shared_ptr<Firestore> firestore_;
+    FSTQuery* query_;
+    QuerySnapshot::Listener user_listener_;
+  };
+  auto view_listener =
+      absl::make_unique<Converter>(this, std::move(user_listener));
+
+  // Call the view_listener on the user Executor.
+  auto async_listener = AsyncEventListener<ViewSnapshot>::Create(
+      firestore_->client().userExecutor, std::move(view_listener));
+
+  std::shared_ptr<QueryListener> query_listener =
+      [firestore_->client() listenToQuery:this->query()
+                                  options:options
+                                 listener:async_listener];
+
+  return ListenerRegistration(firestore_->client(), std::move(async_listener),
+                              std::move(query_listener));
+}
+
+Query Query::Filter(FieldPath field_path,
+                    Filter::Operator op,
+                    FSTFieldValue* field_value,
+                    const std::function<std::string()>& type_describer) const {
+  if (field_path.IsKeyFieldPath()) {
```

Comment:
Not for this PR: this function is somewhat hard to follow (3-level-nested `if` statements in particular).